presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java

```diff
@@ -0,0 +1,296 @@
+                                new NodeId(1, 1),
+                                new NodeId(3, 1),
+                                new NodeId(42, 0))
+                },
+                {ImmutableMap.of(
+                        new NodeId(1, 0), new long[] {1, 2, 3, 4},
+                        new NodeId(3, 0), new long[] {1, 3, 5, 7},
+                        new NodeId(4, 1), new long[] {1, 1, 2, 3},
+                        new NodeId(4, 2), new long[] {2, 4, 6, 8},
+                        new NodeId(4, 4), new long[] {1, 4, 9, 16}),
+                        ImmutableList.of(new NodeId(2, 0),
+                                new NodeId(4, 3),
+                                new NodeId(4, 42),
+                                new NodeId(42, 0))
+                },
+        };
+    }
+
+    @Test(dataProvider = "dataForBufferReuseTest")
+    public void testBufferReuse(long[] buffer, int items, boolean reused)
+            throws IOException
+    {
+        NodeId nodeId = new NodeId(1, 0);
+        long[] data = new long[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16};
+        TestingHiveOrcAggregatedMemoryContext aggregatedMemoryContext = new TestingHiveOrcAggregatedMemoryContext();
+        LongDictionaryProvider dictionaryProvider = new LongDictionaryProvider(
+                createLongDictionaryStreamSources(ImmutableMap.of(nodeId, data), aggregatedMemoryContext),
+                aggregatedMemoryContext.newOrcLocalMemoryContext("testDictionaryProvider"));
+        StreamId streamId = nodeId.toDictionaryDataStreamId();
+        long[] newDictionary = dictionaryProvider.getDictionary(createFlatStreamDescriptor(streamId), buffer, items);
+        assertEquals(newDictionary == buffer, reused);
+        assertTrue(newDictionary.length >= items);
+    }
+
+    @DataProvider(name = "dataForBufferReuseTest")
+    public Object[][] dataForBufferReuseTest()
+    {
+        return new Object[][] {
+                {new long[0], 4, false},
+                {new long[4], 0, true},
+                {new long[4], 4, true},
+                {new long[4], 8, false},
+                {new long[8], 4, true},
+                {new long[16], 12, true},
+                {new long[16], 16, true},
+        };
+    }
+
+    private StreamDescriptor createFlatStreamDescriptor(StreamId streamId)
+    {
+        return new StreamDescriptor("test_dictionary_stream", streamId.getColumn(),
+                String.format("field_%d", streamId.getColumn()), LONG_TYPE, DUMMY_ORC_DATA_SOURCE, ImmutableList.of(), streamId.getSequence());
+    }
+
+    private InputStreamSources createLongDictionaryStreamSources(Map<NodeId, long[]> streams, OrcAggregatedMemoryContext aggregatedMemoryContext)
+    {
+        SharedBuffer decompressionBuffer = new SharedBuffer(aggregatedMemoryContext.newOrcLocalMemoryContext("sharedDecompressionBuffer"));
+        Map dictionaryStreams = new HashMap<StreamId, InputStreamSource<?>>();
```

Comment:
use ImmutableMap.builder()

Reply:
You and @sdruzkin have opposite opinions on this. 😂. @sdruzkin thinks it's a bit wasteful, though I suppose it doesn't matter in tests. @sdruzkin, thoughts on this?

Reply:
Previous comments I got for some of my PRs are ImmutableMaps everywhere even in tests.

Reply:
> ImmutableMaps everywhere even in tests.

That's harsh. In this case we built an ImmutableMap, which was then passed down into a constructor that copied it into another ImmutableMap, which was lame, so I said that using a plain HashMap would be fine.